File: android/src/main/java/com/audiocontext/nodes/audionode/ChannelCountMode.kt
package com.audiocontext.nodes.audionode

enum class ChannelCountMode {
  MAX,
  CLAMPED_MAX,
  EXPLICIT,
  ;

  companion object {
    fun fromString(mode: String): ChannelCountMode =
      when (mode.lowercase()) {
        "max" -> MAX
        "clamped-max" -> CLAMPED_MAX
        "explicit" -> EXPLICIT
        else -> throw IllegalArgumentException("Invalid channel count mode")
      }

    fun toString(mode: ChannelCountMode): String =
      when (mode) {
        MAX -> "max"
        CLAMPED_MAX -> "clamped-max"
        EXPLICIT -> "explicit"
      }
  }
}

File: packages/react-native-audio-api/ios/nodes/AudioNode/ChannelCountMode/ChannelCountMode.m
#import "ChannelCountMode.h"

@implementation ChannelCountMode

+ (ChannelCountModeEnum)channelCountModeFromString:(NSString *)mode
{
  if ([mode isEqualToString:@"max"]) {
    return ChannelCountModeMax;
  }

  if ([mode isEqualToString:@"clamped-max"]) {
    return ChannelCountModeClampedMax;
  }

  if ([mode isEqualToString:@"explicit"]) {
    return ChannelCountModeExplicit;
  }

  @throw
      [NSException exceptionWithName:@"Invalid channel count mode"
                              reason:@"You have to pick correct channel count mode: ['max', 'clamped-max', 'explicit']"
                            userInfo:nil];
}

+ (NSString *)stringFromChannelCountMode:(ChannelCountModeEnum)mode
{
  switch (mode) {
    case ChannelCountModeMax:
      return @"max";
    case ChannelCountModeClampedMax:
      return @"clapmed-max";
    case ChannelCountModeExplicit:
      return @"explicit";
    default:
      @throw [NSException
          exceptionWithName:@"Invalid channel count mode"
                     reason:@"You have to pick correct channel count mode: ['max', 'clamped-max', 'explicit']"
                   userInfo:nil];
  }

  return nil;
}

@end
